### PR TITLE
4.x: Resource limits

### LINKED ISSUE #6819: 4.x: WebServer: provide mechanism to limit number of concurrent requests
To protect against unbounded resource consumption we should consider adding mechanisms to:

* Limit the maximum number of concurrent requests, similar to Apache httpd's [MaxRequestWorkers](https://httpd.apache.org/docs/current/mod/mpm_common.html#maxrequestworkers).  We would need to take into account Http1 connections and Http2 streams.
* Support timeout of idle connects like Apache's [TimeOut](https://httpd.apache.org/docs/2.2/mod/core.html#timeout)

### PR DESCRIPTION
**Pull request by tomas-langer**
Resolves #6819 

Maximal number of TCP connections limit
Maximal number of concurrent requests limit
Idle connection timeout
Align order of parameters in SocketHttpClient (methods sendAndReceive and request)

### CODE REVIEW

File: nima/webserver/webserver/src/main/java/io/helidon/nima/webserver/ServerListener.java [hunk 1]
```diff
@@ -361,16 +383,23 @@ private void listen() {
...
                     readerExecutor.execute(handler);
                 } catch (RejectedExecutionException e) {
                     LOGGER.log(ERROR, "Executor rejected handler for new connection");
+                    // we never started the handler, so we must release the semaphore here
+                    connectionSemaphore.release();
                 } catch (Exception e) {
                     // we may get an SSL handshake errors, which should only fail one socket, not the listener
                     LOGGER.log(TRACE, "Failed to handle accepted socket", e);
+                    // we never started the handler, so we must release the semaphore here
+                    connectionSemaphore.release();
```
Comment (danielkec) [inline nima/webserver/webserver/src/main/java/io/helidon/nima/webserver/ServerListener.java:402]: This covers the case when waiting acquire is interrupted by idle connection timer right?

File: common/http/src/main/java/io/helidon/common/http/Http.java [hunk 2]
```diff
@@ -1890,6 +1883,15 @@ public static ZonedDateTime parse(String text) {
...
         }
 
+        /**
+         * Last recorded timestamp.
+         *
+         * @return timestamp
+         */
+        public static ZonedDateTime timestamp() {
+            return time;
+        }
```
Comment (danielkec) [inline common/http/src/main/java/io/helidon/common/http/Http.java:1893]: Clever!

File: nima/http2/webserver/src/main/java/io/helidon/nima/http2/webserver/Http2Stream.java [hunk 3]
```diff
@@ -384,8 +397,15 @@ private void handle() {
...
+            semaphoreAcquired = requestSemaphore.tryAcquire();
             try {
-                routing.route(ctx, request, response);
+                if (semaphoreAcquired) {
+                    routing.route(ctx, request, response);
+                } else {
+                    ctx.log(LOGGER, TRACE, "Too many concurrent requests, rejecting request.");
+                    response.status(Http.Status.SERVICE_UNAVAILABLE_503)
+                            .send("Too Many Concurrent Requests");
+                }
```
Comment (danielkec) [inline nima/http2/webserver/src/main/java/io/helidon/nima/http2/webserver/Http2Stream.java:408]: We should have test coverage for this too

**Review by danielkec — CHANGES_REQUESTED** [review]

File: nima/http2/webserver/src/main/java/io/helidon/nima/http2/webserver/Http2Stream.java (hunk 3 above)
Comment (tomas-langer) [reply nima/http2/webserver/src/main/java/io/helidon/nima/http2/webserver/Http2Stream.java:408]: There is test coverage for all three, 503 in io.helidon.tests.integration.webserver.resourcelimit.MaxConcurrentRequestsTest

Comment (tomas-langer) [reply nima/http2/webserver/src/main/java/io/helidon/nima/http2/webserver/Http2Stream.java:408]: ah, I see, for HTTP/2

File: nima/webserver/webserver/src/main/java/io/helidon/nima/webserver/ServerListener.java (hunk 1 above)
Comment (tomas-langer) [reply nima/webserver/webserver/src/main/java/io/helidon/nima/webserver/ServerListener.java:402]: yes it should, as the exception would be caught here. I have tried to cover all "escape routes" from the code, and always release the semaphore.

File: nima/http2/webserver/src/main/java/io/helidon/nima/http2/webserver/Http2Stream.java (hunk 3 above)
Comment (tomas-langer) [reply nima/http2/webserver/src/main/java/io/helidon/nima/http2/webserver/Http2Stream.java]: Fixed, added HTTP/2 support for testing, added test.

File: common/http/src/main/java/io/helidon/common/http/Http.java (hunk 2 above)
Comment (tomas-langer) [reply common/http/src/main/java/io/helidon/common/http/Http.java:1893]: Wanted to keep the performance impact to a minimum, and we do not care about seconds that much in this case.

File: nima/testing/junit5/http2/src/main/java/io/helidon/nima/testing/junit5/http2/Http2ServerExtension.java [hunk 4]
```diff
@@ -0,0 +1,59 @@
...
+    @Override
+    public Object resolveParameter(ParameterContext parameterContext,
+                                   ExtensionContext extensionContext,
+                                   Class<?> parameterType,
+                                   WebServer server) {
+        String socketName = Junit5Util.socketName(parameterContext.getParameter());
+
+        if (Http2Client.class.equals(parameterType)) {
+            return Http2Client.builder()
+                    .baseUri("ws://localhost:" + server.port(socketName))
```
Comment (danielkec) [inline nima/testing/junit5/http2/src/main/java/io/helidon/nima/testing/junit5/http2/Http2ServerExtension.java:54]: ws scheme?

Comment (tomas-langer) [reply nima/testing/junit5/http2/src/main/java/io/helidon/nima/testing/junit5/http2/Http2ServerExtension.java:54]: ouch

**Review by danielkec — APPROVED** [review]

**Pull request merged**